Remove all calls to the /images endpoint

jwforres opened the pull request.

jwforres commented:
@liggitt review

Currently the caching of images by data service isn't working as expected because the "name" we get back from the API doesn't match the "name" that we request.  Waiting for a decision from @ncdc and @smarterclayton on this.   If we change the Kind to "ImageStreamImage" and the name to what we actually requested with, then no changes are needed in the console.  If the api is going to keep returning Kind of "Image" and the name as the image's id then the console will need to specially handle how it caches imageStreamImage requests.

jwforres commented:
Per the discussion here https://github.com/openshift/origin/issues/1701#issuecomment-92350617  the Kind will be corrected, so the caching should work as written.

liggitt commented on assets/app/scripts/controllers/overview.js:
```diff
@@ -231,14 +233,43 @@ angular.module('openshiftConsole')
+      $scope.imageStreams = imageStreams.by("metadata.name");
+      buildDockerRefMapForImageStreams();
+      fetchReferencedImages();
+      // TODO look for "referenced images"
```
remove TODO? isn't that what `fetchReferencedImages` does?

liggitt commented on assets/app/scripts/controllers/overview.js:
```diff
@@ -15,8 +15,9 @@ angular.module('openshiftConsole')
-    $scope.images = {};
+    $scope.imageStreams = {};
     $scope.imagesByDockerReference = {};
+    $scope.imageStreamImageRefByDockerReference = {};
```
comment here? "imageStreamImageRefByDockerReference lets us determine if a particular container's docker image reference belongs to an imageStream"

liggitt commented on assets/app/scripts/controllers/overview.js:
```diff
@@ -231,14 +233,43 @@ angular.module('openshiftConsole')
+      Logger.log("imageStreams (subscribe)", $scope.imageStreams);
     }));
 
+    var fetchReferencedImages = function() {
```
`fetchReferencedImageStreamImages`? Any way to reuse this and `buildDockerRefMapForImageStreams` between the overview and pods controllers?

liggitt replied:
since this is asynchronous, make it return a promise that gets fulfilled when the fetching is done, in case anything needs to do follow-up work

liggitt commented on assets/app/scripts/controllers/overview.js:
```diff
@@ -231,14 +233,43 @@ angular.module('openshiftConsole')
+      angular.forEach($scope.pods, function(pod){
+        angular.forEach(pod.spec.containers, function(container){
+          var dockerRef = container.image;
+          if (!$scope.imagesByDockerReference[dockerRef] && $scope.imageStreamImageRefByDockerReference[dockerRef]) {
```
unnesting and commenting would be helpful...
```
var dockerRef = container.image;
if ($scope.imagesByDockerReference[dockerRef]) {
  // Already have an image for this ref
  return;
}

var imageStreamImageRef = $scope.imageStreamImageRefByDockerReference[dockerRef];
if (!imageStreamImageRef) {
  // This docker ref doesn't come from an image stream we know about
  return
}

DataService.get("imageStreamImages", imageStreamImageRef, $scope).then(function(image) {
    $scope.imagesByDockerReference[dockerRef] = image;
});
```

liggitt commented on assets/app/scripts/controllers/overview.js:
```diff
@@ -231,14 +233,43 @@ angular.module('openshiftConsole')
+        angular.forEach(imageStream.status.tags, function(tag){
+          angular.forEach(tag.items, function(image){
+            if (image.image) {
+              $scope.imageStreamImageRefByDockerReference[image.dockerImageReference] = imageStream.metadata.name + "@" + image.image;
```
seems a little weird to have to build this ourselves...

jwforres replied:
Meaning you think the imagestreamName@imageId should be returned on the 
items under the tag?  CC @ncdc

On 04/13/2015 10:51 AM, Jordan Liggitt wrote:
>
> In assets/app/scripts/controllers/overview.js 
> <https://github.com/openshift/origin/pull/1699#discussion_r28244272>:
>
> > +            DataService.get("imageStreamImages", $scope.imageStreamImageRefByDockerReference[container.image], $scope).then(
> > +              function(image) {
> > +                $scope.imagesByDockerReference[dockerRef] = image;
> > +              }
> > +            );
> > +          }
> > +        });
> > +      })
> > +    };
> > +
> > +    var buildDockerRefMapForImageStreams = function() {
> > +      angular.forEach($scope.imageStreams, function(imageStream) {
> > +        angular.forEach(imageStream.status.tags, function(tag){
> > +          angular.forEach(tag.items, function(image){
> > +            if (image.image) {
> > +              $scope.imageStreamImageRefByDockerReference[image.dockerImageReference] = imageStream.metadata.name + "@" + image.image;
>
> seems a little weird to have to build this ourselves...
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/openshift/origin/pull/1699/files#r28244272>.
>

liggitt replied:
Yeah, making clients concatenate stuff themselves isn't good... better to give them an opaque value they can directly compare for equality

ncdc replied:
You shouldn't need to do this (or check `image.image`). Just use `image.dockerImageReference` as your pull spec. Or am I misunderstanding what this is used for?

liggitt replied:
We're trying to correlate imagestream images with specific pod.spec.containers[N].image docker image references...

ncdc replied:
Also, `image.image` may be a v1 "fake by-id id" (actually a tag), in which case trying to use `<stream>@<image.image>` won't work.

ncdc replied:
> We're trying to correlate imagestream images with specific pod.spec.containers[N].image docker image references

Yes, definitely just use `image.dockerImageReference` for that.

liggitt replied:
Mmm... I misread. The key of our map is the `pod.spec.containers[N].image` value... we're already using `image.dockerImageReference` for that.

The value of the map is the name we will use to call `GET /osapi/.../imageStreamImages/[NAME]`.

ncdc replied:
In that case, you'll need to concat. We do provide an opaque data point, but it's the docker pull spec (dockerImageReference), not the `imageStreamImage` reference.

liggitt replied:
Why doesn't TagEvent.Image contain the full name you would use to GET imageStreamImages?

ncdc replied:
That's the image's "id". Images are not namespaced nor do they belong to image streams (to support deduplication), but only the registry is allowed to retrieve an image solely by its id. Users must retrieve images using imageStreamImages where they specify namespace + image stream + id.

I can't say I ever previously considered having TagEvent.Image be of the form `<namespace>/<image stream>@<id>`. @smarterclayton WDYT?

liggitt replied:
So what we're looking for is an `imageStreamImageReference`, I guess? Seems like making TagEvent.Image serve both purposes (identify the global image name *and* the `@...` portion of the imageStreamImage name) is a bad idea... should TagEvent provide three bits: the docker image reference, the image name, and the imagestreamimage name?

liggitt replied:
@ncdc could we do something like #1710 to avoid making API consumers know how to construct imageStreamImage names?

jwforres commented:
@liggitt addressed all feedback other then the ongoing discussion around whether the API will return the ImageStreamImage name, can always switch to use that in a follow-up

liggitt commented on assets/app/scripts/controllers/pods.js:
```diff
@@ -29,13 +31,14 @@ angular.module('openshiftConsole')
+      $scope.imageStreams = imageStreams.by("metadata.name");
+      ImageStreamResolver.buildDockerRefMapForImageStreams($scope.imageStreams, $scope.imageStreamImageRefByDockerReference);
+      ImageStreamResolver.fetchReferencedImageStreamImages($scope.pods, $scope.imagesByDockerReference, $scope.imageStreamImageRefByDockerReference, $scope);
+      // TODO look for "referenced images"
```
is this TODO still relevant?

jwforres replied:
no it got copied, didnt realize it was in two places

liggitt commented on assets/app/scripts/services/imageStreamResolver.js:
```diff
@@ -0,0 +1,54 @@
+      angular.forEach(imageStream.status.tags, function(tag){
+        angular.forEach(tag.items, function(image){
+          if (image.image) {
+            imageStreamImageRefByDockerReference[image.dockerImageReference] = imageStream.metadata.name + "@" + image.image;
```
add a TODO to use built value from API if/when available

liggitt commented on assets/app/scripts/services/imageStreamResolver.js:
```diff
@@ -0,0 +1,54 @@
+    });
+  };
+
+  return new ImageStreamResolver();;
```
`;;`

liggitt commented:
nits, LGTM. merge once the kind stuff is fixed well enough for this to work

jwforres commented:
works as is, caching will just improve with the Kind changes  [merge]
